engine/table/src/main/java/io/deephaven/engine/table/impl/UngroupOperation.java

```diff
@@ -431,214 +332,420 @@ private class UngroupListener extends BaseTable.ListenerImpl {
+         * @param modifiedPreShift the modified index, in pre-shift space
+         * @param addedBuilder the builder for rows added by modifications
+         * @param removedBuilder the builder for rows removed by modifications
+         * @param modifyBuilder the builder for rows modified by modifications
+         * @param currentSizes an output parameter when we determine a change of base is necessary; if we've already
+         *        computed these values we stuff them in the mutable, so that they can be reused as input again
+         * @return the necessary base
+         */
         private int evaluateModified(final RowSet modified,
                 final RowSet modifiedPreShift,
-                final RowSetBuilderSequential modifyBuilder,
-                final RowSetBuilderSequential addedBuilded,
+                final RowSetBuilderSequential addedBuilder,
                 final RowSetBuilderSequential removedBuilder,
-                final int currentBase) {
+                final RowSetBuilderSequential modifyBuilder,
+                final int base,
+                final MutableObject<long[]> currentSizes) {
             if (modified.isEmpty()) {
-                return currentBase;
+                return base;
             }
 
-            final long base = shiftState.getNumShiftBits();
             final int size = modified.intSize("ungroup");
-            final long[] sizes = new long[size];
-            final long[] prevSizes = new long[size];
-
-            final long maxSize = computeMaxSize(modified, sizes, false);
-            computePrevSize(modifiedPreShift, prevSizes);
-            final int minBase = 64 - Long.numberOfLeadingZeros(maxSize);
-            if (minBase > currentBase) {
-                // If we are going to force a rebase, there is no need to compute the entire index
-                return minBase;
+            final long[] sizes;
+            final long maxSize;
+            if (currentSizes.get() == null) {
+                sizes = new long[size];
+                maxSize = computeMaxSize(modified, sizes, false);
+                final int minBase = determineRequiredBase(maxSize);
+                if (minBase > base) {
+                    // If we are going to force a rebase, there is no need to compute the entire rowset, but the caller
+                    // would
+                    // like our copy of the sizes we computed
+                    currentSizes.setValue(sizes);
+                    return minBase;
+                }
+            } else {
+                sizes = currentSizes.get();
+                maxSize = Arrays.stream(sizes).max().getAsLong();
+                final int minBase = determineRequiredBase(maxSize);
+                Assert.leq(minBase, "minBase", minBase, "maxBase");
             }
 
+
+            final long[] prevSizes = new long[size];
+            computePrevSize(result.getRowSet().prev(), shiftState.getPrevNumShiftBits(), modifiedPreShift, prevSizes);
```

Comment:
I find it confusing to look up `shiftState.getPrevNumShiftBits()`, because in all cases that will be equal to `shiftState.getNumShiftBits()`.